// @react-three/drei@9.102.0/core/Edges.js
import _extends from '@babel/runtime/helpers/esm/extends';
import * as React from 'react';
import * as THREE from 'three';
import { LineSegmentsGeometry, LineMaterial } from 'three-stdlib';
import { useThree } from '@react-three/fiber';

const Edges = /* @__PURE__ */React.forwardRef(({
  children,
  threshold = 15,
  color = 'black',
  opacity = 1,
  linewidth,
  lineWidth,
  ...props
}, fref) => {
  var _ref;
  const ref = React.useRef(null);
  React.useImperativeHandle(fref, () => ref.current, []);
  const geometry = React.useMemo(() => new LineSegmentsGeometry(), []);
  const material = React.useMemo(() => new LineMaterial(), []);
  const size = useThree(s => s.size);
  const memoizedGeometry = React.useRef();
  const memoizedThreshold = React.useRef();
  React.useLayoutEffect(() => {
    const parent = ref.current.parent;
    if (parent) {
      const geometry = parent.geometry;
      if (geometry !== memoizedGeometry.current || threshold !== memoizedThreshold.current) {
        memoizedGeometry.current = geometry;
        memoizedThreshold.current = threshold;
        const points = new THREE.EdgesGeometry(geometry, threshold).attributes.position.array;
        ref.current.geometry.setPositions(points);
        ref.current.geometry.attributes.instanceStart.needsUpdate = true;
        ref.current.geometry.attributes.instanceEnd.needsUpdate = true;
      }
    }
  });
  return /*#__PURE__*/React.createElement("mesh", _extends({
    ref: ref,
    raycast: () => null
  }, props), /*#__PURE__*/React.createElement("primitive", {
    object: geometry,
    attach: "geometry"
  }), /*#__PURE__*/React.createElement("primitive", {
    object: material,
    attach: "material",
    color: color,
    opacity: opacity,
    transparent: opacity < 1,
    resolution: [size.width, size.height],
    linewidth: (_ref = linewidth !== null && linewidth !== void 0 ? linewidth : lineWidth) !== null && _ref !== void 0 ? _ref : 1
  }), children);
});

export { Edges };
